### DISCUSSION TITLE
Unix timestamp in HLS stream

### DISCUSSION THREAD

**Discussion by giano574** [bluenviron/mediamtx#1135] [Questions]
Is there a way to add the unix timestamp to each frame in a HLS stream as meta data? 
Alternatively, can the PTS be changed to be unix timestamps?

**Comment by giano574** [comment]
Can the `pcrOffset` in [writer.go:15](https://github.com/aler9/rtsp-simple-server/blob/main/internal/hls/mpegts/writer.go#L15) be used for this? Perhaps adding a setting to the configuration called `realTimePtsAndDts` or similar, that if true sets the `pcrOffset` to the unix timestamp of the first frame?

Alternatively, in FFmpeg (which I use to read the stream), the `AVFormatContext` struct has a field called [start_time_realtime](https://ffmpeg.org/doxygen/4.4/structAVFormatContext.html#aa5ddb5cee1df28f21739133f2e37f1c5) that is currently set to `AV_NOPTS_VALUE ` when reading a HLS stream published with rtsp-simple-server. I don't know how FFmpeg sets the field but if that could be set from additional data generated by rtsp-simple-server that would be equally useful

**Comment by aler9** [comment]
Hello,
The absolute timestamp of each video frame (i.e. the RTC timestamp) is already available inside HLS streams generated by rtsp-simple-server. In order to extract it, you have to start from the RTC timestamp of each segment, that is saved into the `EXT-X-PROGRAM-DATE-TIME` field:

```
#EXT-X-PROGRAM-DATE-TIME:2022-10-25T13:15:46.953Z
#EXTINF:8.333333333,
seg0.ts
```

Then, the RTC timestamp of each frame is 

```
time_frame = program_date_time_of_segment + player_position - player_position_when_segment_was_loaded
```

If you're using `hls.js`, the `EXT-X-PROGRAM-DATE-TIME` value can be extracted during the `FRAG_CHANGED` event (`data.frag.rawProgramDateTime`). You also have to extract the player position (`videoElement.currentTime`):

```js

engine.on(HlsJs.Events.FRAG_CHANGED, (evt, data) => {
    if (data.frag.rawProgramDateTime === null) {
        return;
    }

    this.lastFragmentDate = (new Date(data.frag.rawProgramDateTime)).getTime();
    this.lastCurrentTime = videoElement.currentTime;
}
```

Then, you can launch a routine that prints the exact RTC timestamp of the current frame:

```js
window.setInterval(() => {
    curDate = this.lastFragmentDate + (videoElement.currentTime - this.lastCurrentTime * 1000);
}, 50);
```

the `1000` factor is needed since `lastFragmentDate` is in milliseconds, while `currentTime` is in seconds.

**Reply by giano574** [reply]
Unfortunately, I'm stuck with using the FFmpeg API for reading the stream, and I haven't been able to find any way to extract EXT-X-PROGRAM-DATE-TIME from the current segment. In fact, I can't even seem to figure out which segment is currently playing.

**Reply by aler9** [reply]
If you want to use the FFmpeg API, in my opinion you'd better use the RTSP protocol to pull the stream from the server. Latency is way lower and `start_time_realtime` is populated correctly:
https://github.com/FFmpeg/FFmpeg/blob/2b41463b8706638b9aaf967655f95ddc195c32eb/libavformat/rtsp.c#L2314

**Reply by giano574** [reply]
Unfortunately I can't switch to RTSP, since the network connection is not stable enough for that. Great suggestion, though.

[a comment by an automated account is not shown]
